src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/ProxyWhitelist.java

```diff
@@ -122,77 +155,168 @@ public ProxyWhitelist(Whitelist... delegates) {
 
     @Override public final boolean permitsMethod(Method method, Object receiver, Object[] args) {
         synchronized (this.delegates) {
-            for (Whitelist delegate : delegates) {
+            String sigString = Whitelist.canonicalMethodSig(method);
+            Boolean b = permittedCache.get(sigString);
```

Comment:
Is this really enough of a check? Only the method signature?. The actual ask is: is it allowed to run this method on __this__ object with __these__ arguments.

See for example [this implementation](https://github.com/jenkinsci/email-ext-plugin/blob/master/src/main/java/hudson/plugins/emailext/groovy/sandbox/MimeMessageInstanceWhitelist.java) for a bit unorthodox implementation.

Reply:
@rsandell So... here we're relying on a couple properties of the Whitelist.  First, for Whitelists implementing the `CacheableWhitelist` interface, they MUST only depend on the method/field being used and NOT on the object or arguments used.  Second, any Whitelist that permits the method means it can be used.  The combination means that we can cache an approval *if* it comes from a CacheableWhitelist. 

If the call/access is *not* permitted or was permitted by a non-Cacheable Whitelist, we know it doesn't come from a Cacheable whitelist and have to dynamically check it with each invocation, because the approval depends on the receiver or arguments.  But wait!  We can skip the CacheableWhitelist checks because know at that point that none of them granted permission, so *none of them are going to.* 

Make a bit more sense?